### PR TITLE
Added threading to the internal downloader

### PR DESCRIPTION
**Pull request by Blatzar**
The downloading should be full working.

Current ~~issues~~ todos: 
Creates a part file with zeroes making any cancellation and resumption say it's fully downloaded.
Max_threads not yet in config.
m3u8 downloader using this.

For full disclosure I copied parts from https://www.geeksforgeeks.org/simple-multithreaded-download-manager-in-python/

### CODE REVIEW

**Comment by Blatzar** [comment]
Error: turns out the progress report is currently incorrent

**Comment by AbdullahM0hamed** [comment]
Do you mean 'incoherent'?

**Comment by Blatzar** [comment]
BUG: somehow the files generated on windows from this internal downloader doesn't work, not playable by mpv or vlc.
Moving the linux generated mp4 to windows works so it's very likely the writing is at fault.

File: anime_downloader/downloader/base_downloader.py [hunk 1]
```diff
@@ -28,25 +28,35 @@ def __init__(self, source, path, force, range_size, callback=None):
...
-        r = requests.get(
-            self.source.stream_url, headers=headers, stream=True)
-
-        self._total_size = int(r.headers['Content-length'])
-        logger.debug('total size: ' + str(self._total_size))
-        if os.path.exists(self.path):
-            if abs(os.stat(self.path).st_size - self._total_size) < 10 \
-               and not self.force:
-                logger.warning('File already downloaded. Skipping download.')
+        for i in range(5):
```
Comment (vn-ki) [inline anime_downloader/downloader/base_downloader.py:38]: What does this do?

Comment (Blatzar) [reply anime_downloader/downloader/base_downloader.py:38]: Just simple retries if the first request somehow failed

File: anime_downloader/downloader/base_downloader.py [hunk 2]
```diff
@@ -17,36 +15,50 @@ def __init__(self, source, path, force, range_size, callback=None):
...
-
+        self.resumed = 0
         # these should be included in a options dict, maybe
         self.force = force
         self.range_size = range_size
 
         if callback is None:
             callback = write_status
         self.callback = callback
+        self.thread_report = []
```
Comment (vn-ki) [inline anime_downloader/downloader/base_downloader.py:26]: Thread code shouldn't be in base downloader

File: anime_downloader/extractors/base_extractor.py [hunk 3]
```diff
@@ -3,15 +3,15 @@
 
 
 class BaseExtractor:
-    def __init__(self, url, quality=None, headers=None):
+    def __init__(self, url, quality=None, headers={}):
```
Comment (vn-ki) [inline anime_downloader/extractors/base_extractor.py:6]: Do not pass `{}` in default arguments.

File: anime_downloader/util.py [hunk 4]
```diff
@@ -274,13 +273,19 @@ def format_command(cmd, episode, file_format, speed_limit, path):
...
+    else:
+        useragent = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.87 Safari/537.36'
 
     rep_dict = {
         'stream_url': episode.source().stream_url if not episode.url.startswith('magnet:?xt=urn:btih:') else episode.url,
         'file_format': file_format,
         'download_dir': os.path.abspath(path),
         'referer': episode.source().referer,
-        'useragent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.87 Safari/537.36',
+        'useragent': f'"{useragent}"',
```
Comment (vn-ki) [inline anime_downloader/util.py:288]:
```suggestion
        'useragent': useragent,
```
?

**Comment by Blatzar** [comment]
Closing as superseded by PySmartDl

**Comment by ArjixGamer** [comment]
plz dont remove the branch

**Comment by Blatzar** [comment]
I'll merge it to another branch for savekeeping

**Pull request merged**